aten/src/ATen/core/dispatch/KernelRegistration.h

```diff
@@ -49,17 +45,97 @@ class KernelRegistrar final {
 
   ~KernelRegistrar() {
     if (owns_registration_) {
-      Dispatcher<OpSchemaDef>::deregisterKernel(dispatch_key_);
+      Dispatcher::singleton().deregisterKernel(op_(), dispatch_key_);
     }
   }
 
 private:
-  const typename Schema::dispatch::dispatch_key_type dispatch_key_;
+  const OperatorHandle& (*op_)();
+  const TensorTypeId dispatch_key_;
   bool owns_registration_;
 
   C10_DISABLE_COPY_AND_ASSIGN(KernelRegistrar);
 };
 
+namespace detail {
+template<class T>
+struct ivalue_to_arg_type {
+  static T call(const IValue& v) {
+    return std::move(v).to<T>();
+  }
+};
+template<class T>
+struct ivalue_to_arg_type<ArrayRef<T>> {
+  static ArrayRef<T> call(const IValue& v) {
+    return v.to<intrusive_ptr<ivalue::List<T>>>()->elements();
+  }
+};
+
+template<class FuncType, FuncType* func, class... ExtraArgs, size_t... ivalue_arg_indices>
```

Comment:
I know you don't like documenting internal functions, but even just a sentence for these gnarly template metaprogramming functions will help people who don't know very much about C++ templates *a lot*